src/Storages/MergeTree/MergeTreeSink.cpp

```diff
@@ -96,17 +103,15 @@ void MergeTreeSink::consume(Chunk chunk)
         if (!support_parallel_write && temp_part.part->getDataPartStorage().supportParallelWrite())
             support_parallel_write = true;
 
-        String block_dedup_token;
-        if (storage.getDeduplicationLog())
+        String block_id;
+        if (deduplicate)
         {
-            const String & dedup_token = settings.insert_deduplication_token;
-            if (!dedup_token.empty())
-            {
-                /// multiple blocks can be inserted within the same insert query
-                /// an ordinal number is added to dedup token to generate a distinctive block id for each block
-                block_dedup_token = fmt::format("{}_{}", dedup_token, chunk_dedup_seqnum);
-                ++chunk_dedup_seqnum;
```

Comment:
The problem was because of this line. A multithread `INSERT SELECT` creates multiple `sinks` and each one of those had its own `chunk_dedup_seqnum` counter without any synchronization between them. As a result some of the inserted blocks were incorrectly considered as "duplicates".